File: pp-exercises/mpi/exercise1/master-worker/master-worker.c
# include <stdio.h>
# include <stdlib.h>
# include <mpi.h>

void master(int);
void worker(void);

//------------------------------------------------------------------------------
int main(int argc, char *argv[])   // master-worker
{
    int maxtask = 40;
    int rank;

    MPI_Init(&argc, &argv);

    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    if (rank == 0)
	master(maxtask);
    else
	worker();
    
    MPI_Finalize();
    return 0;
}

//------------------------------------------------------------------------------
void master(const int maxtask)
{
    MPI_Status status;
    int size, msg, task, dest;

    MPI_Comm_size(MPI_COMM_WORLD, &size);

    for (task = 1; task <= maxtask; task++)
    {
	MPI_Recv(&msg, 0, MPI_INT, MPI_ANY_SOURCE, 0, MPI_COMM_WORLD, &status);
        MPI_Send(&task, 1, MPI_INT, status.MPI_SOURCE, 0, MPI_COMM_WORLD);
    }

    task = -1;
    for (dest = 1; dest < size; dest++)
    {
	MPI_Recv(&msg, 0, MPI_INT, dest, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
	MPI_Send(&task, 1, MPI_INT, dest, 0, MPI_COMM_WORLD);
    }
}

//------------------------------------------------------------------------------
void worker(void)
{
    int rank, msg, task;

    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    do
    {
	MPI_Send(&msg, 0, MPI_INT, 0, 0, MPI_COMM_WORLD);
	MPI_Recv(&task, 1, MPI_INT, 0, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);

	if (task >= 0)
	{
	    printf("rank %d: working on task %d\n", rank, task);
	    system("sleep 1");
	}
    }
    while (task >= 0);
}

File: pp-exercises/mpi/exercise1/master-worker/master-worker.f90
!==============================================================================
program main  ! master-worker

  implicit none
  include 'mpif.h'
  integer, parameter :: maxtask = 40
  integer :: rank, ierr
 
  call mpi_init(ierr)

  call mpi_comm_rank(MPI_COMM_WORLD, rank, ierr)

  if (rank == 0) then
     call master(maxtask)
  else
     call worker()
  endif

  call mpi_finalize(ierr)
end

!------------------------------------------------------------------------------
subroutine master(maxtask)

  implicit none
  include 'mpif.h'
  integer, intent(in) :: maxtask
  integer :: size, status(MPI_STATUS_SIZE), ierr
  integer :: msg, task, dest

  call mpi_comm_size(MPI_COMM_WORLD, size, ierr)

  do task = 1, maxtask
     call mpi_recv(msg, 0, MPI_INTEGER, MPI_ANY_SOURCE, 0, MPI_COMM_WORLD, status, ierr)
     call mpi_send(task, 1, MPI_INTEGER, status(MPI_SOURCE), 0, MPI_COMM_WORLD, ierr)
  enddo

  task = -1
  do dest = 1, size - 1
     call mpi_recv(msg, 0, MPI_INTEGER, dest, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)
     call mpi_send(task, 1, MPI_INTEGER, dest, 0, MPI_COMM_WORLD, ierr)
  enddo
end

!------------------------------------------------------------------------------
subroutine worker()

  implicit none
  include 'mpif.h'
  integer :: rank, ierr
  integer :: msg, task

  call mpi_comm_rank(MPI_COMM_WORLD, rank, ierr)

  do 
     call mpi_send(msg, 0, MPI_INTEGER, 0, 0, MPI_COMM_WORLD, ierr)
     call mpi_recv(task, 1, MPI_INTEGER, 0, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr) 

     if (task < 0) exit
     write(6,*) "rank " , rank, ": working on task ", task
     call system("sleep 1")
  enddo

end

!==============================================================================
